### ISSUE TITLE
[DRAFT] Support Rewarder Display

### ISSUE THREAD

**Issue by YuriyVelichkoPI** [prebid/prebid-mobile-ios#713]
## Motivation

The rewarded ads are not just about video format.  But Prebid SDK supports only Rewarded Video Ads.

We should introduce support for Rewarded Display ads as well. 

## Description

// TODO

## Objectives

// TODO

**Comment by YuriyVelichkoPI** [comment]
Yes. SDK has basic support for Rewarded Video ads, but not Display (banner) ones.

**Comment by mdanylov-sigma** [comment]
# PART I — iOS IMPLEMENTATION (PR #1274)

## 1. Problem Statement (iOS)

Issue #713 notes that the Prebid Mobile iOS SDK only supports Rewarded Video ads. Rewarded inventory is not exclusively video — ad servers and mediation platforms also support rewarding users for viewing a display (banner/HTML) creative, e.g. a full-screen MRAID interstitial with a timed reward instead of a video completion event. Publishers currently have no supported way to:

- Request a rewarded ad with a banner/HTML creative format through the Original API.
- Request a rewarded ad with a banner format, a video format, or both, through the Mediation API (AdMob, MAX, and other mediation adapters).
- Get correct fullscreen/interstitial request signaling for a rewarded ad built through the multiformat (in-app bidding) `PrebidAdUnit` API when the publisher only sets `isRewarded` and not `isInterstitial`.

This proposal describes the API surface added in PR #1274 to close those gaps, and asks the community to confirm the shape before the PR comes out of draft.

## 2. Background: What Already Exists

Rewarded ads reach the screen through four distinct surfaces in this SDK, and they don't all render the same way. That distinction matters for this proposal, so it's worth being precise before describing the change:

- **Original API (`RewardedVideoAdUnit`, and now `RewardedDisplayAdUnit`) is ad-server-rendered, not Prebid-rendered.** Prebid's role is limited to `fetchDemand(...)`: building the bid request and attaching `hb_*` targeting to the publisher's ad request object. In the Original API demo this PR adds (`Examples/GAM/OriginalAPI`), the creative is displayed by GAM's own native `RewardedAd` format, and GAM's SDK — not Prebid — decides when the reward is earned and delivers it via its own callback (e.g. `GADFullScreenPresentingAd`'s `userDidEarnRewardHandler`). Prebid's reward timer and close-button logic never run on this path. (AdMob and MAX are not part of the Original API surface here — the AdMob and MAX demo controllers this PR adds use the Mediation API, described next.)
- **Mediation API (`MediationRewardedAdUnit`, demoed here with AdMob and MAX) and the pre-existing GAM Rendering API (`RewardedAdUnit` / `BaseRewardedAdUnit`) are Prebid-rendered.** When Prebid's bid wins the ad network's mediation waterfall (via the Prebid-authored adapter, e.g. `PrebidAdMobRewardedAdapter` in the `prebid-mobile-ios-adapters` repo) or the GAM line item (via `GAMRewardedAdEventHandler`), Prebid builds its own `InterstitialController` from the winning bid and renders the creative itself. It's only on these two surfaces that the close-button suppression, reward timer, and close-action handling in `PBMHTMLCreative.setupRewardTimerIfNeeded(...)` and `ModalViewController` (`creativeDisplayCompleted`, `setupCloseButtonVisibility`) actually run — both key off `adConfiguration.isRewarded`, not off creative format, so this logic already worked for banner/HTML creatives before this PR. The GAM Rendering API row already supported rewarded display before PR #1274 for exactly this reason; it isn't part of this diff.
- **The gap PR #1274 closes is entirely in the public AdUnit configuration surface for Original API and Mediation API.** Nothing today lets a publisher declare "rewarded + banner format" through Original API, or "rewarded + banner and/or video" through Mediation API. That is what this proposal adds — see §4 for the full surface-by-surface breakdown.

Because the Prebid-rendered paths' reward mechanics were already generic, this proposal is intentionally narrow: for Original API and Mediation API it only touches ad-unit configuration classes, not the creative/reward runtime.

## 3. Proposed API

### 3.1 Original API — new `RewardedDisplayAdUnit`

A new public class, sibling to `RewardedVideoAdUnit`, subclassing `InterstitialAdUnit` (which already defaults to banner format and fullscreen position):

```swift
@objcMembers
public class RewardedDisplayAdUnit: InterstitialAdUnit {

    public override init(configId: String) {
        super.init(configId: configId)
        adUnitConfig.adConfiguration.isRewarded = true
    }

    public convenience init(
        configId: String,
        minWidthPerc: Int,
        minHeightPerc: Int
    ) {
        self.init(configId: configId)
        adUnitConfig.minSizePerc = NSValue(
            cgSize: CGSize(width: minWidthPerc, height: minHeightPerc)
        )
    }
}
```

Usage mirrors `RewardedVideoAdUnit`:

```swift
let adUnit = RewardedDisplayAdUnit(configId: storedImpId)
adUnit.bannerParameters.api = [.MRAID_2]
adUnit.bannerParameters.adSizes = [CGSize(width: 320, height: 480)]
adUnit.fetchDemand(adObject: gamRequest) { resultCode in
    // load & present via GAM as usual
}
```

### 3.2 Mediation API — mutable `adFormats` on `MediationRewardedAdUnit`

`MediationRewardedAdUnit` previously hardcoded `adFormats = [.video]` with no way to override it. This proposal makes it a read/write property and adds a convenience initializer:

```swift
@objcMembers
public class MediationRewardedAdUnit: MediationBaseInterstitialAdUnit {

    public var adFormats: Set<AdFormat> {
        get { adUnitConfig.adFormats }
        set { adUnitConfig.adFormats = newValue }
    }

    public init(configId: String, mediationDelegate: PrebidMediationDelegate) {
        // unchanged: defaults to isRewarded = true, adFormats = [.video]
        // (adFormats is now a mutable property, but this initializer's
        // default stays [.video] so existing callers are unaffected)
    }

    public convenience init(
        configId: String,
        mediationDelegate: PrebidMediationDelegate,
        adFormats: Set<AdFormat>
    ) {
        self.init(configId: configId, mediationDelegate: mediationDelegate)
        self.adFormats = adFormats
    }
}
```

Publishers pick one of three shapes:

- `adFormats: [.banner]` — display rewarded
- `adFormats: [.video]` — today's default, unchanged behavior
- `adFormats: [.banner, .video]` — multiformat rewarded; the server chooses the winning format per the standard multiformat auction rules already used elsewhere in the SDK

`bannerParameters` and `videoParameters` both remain accessible regardless of the current `adFormats` value, so a publisher can pre-configure both before or after switching formats.

### 3.3 In-App Bidding (Multiformat) — fullscreen inference fix

`PrebidAdUnit.fetchDemand(request:)` builds an `AdUnitConfig` from a `PrebidRequest`. Previously it copied `isInterstitial` verbatim onto `isInterstitialAd`, and `isRewarded` separately, with no interaction between the two:

```swift
// before
adUnit.adUnitConfig.adConfiguration.isInterstitialAd = request.isInterstitial
adUnit.adUnitConfig.adConfiguration.isRewarded = request.isRewarded
```

A publisher who set `isRewarded: true` but left `isInterstitial` at its default (`false`) would get a rewarded bid request that wasn't flagged as fullscreen/interstitial — inconsistent with every other rewarded surface in the SDK, where rewarded is unconditionally fullscreen. The fix:

```swift
// after
// Rewarded inventory is fullscreen/interstitial even when
// request.isInterstitial is not explicitly set.
adUnit.adUnitConfig.adConfiguration.isInterstitialAd =
    request.isInterstitial || request.isRewarded
adUnit.adUnitConfig.adConfiguration.isRewarded = request.isRewarded
```

### 3.4 Collateral fix — `RewardedVideoAdUnit` (Original API) wasn't declaring `isRewarded` on the request

While implementing §3.1, we found that `RewardedVideoAdUnit`'s initializer never marked its `AdUnitConfig` as rewarded before building the bid request. So the outgoing OpenRTB request could omit the rewarded-inventory signal (`imp.rwdd`, plus the legacy `ext.prebid.is_rewarded_inventory` extension) that Prebid Server uses for rewarded-eligible demand filtering — even though the class exists specifically to request rewarded video. This proposal adds the missing line:

```swift
public init(configId: String) {
    super.init(configId: configId, size: nil, adFormats: [.video])
    adUnitConfig.adConfiguration.isInterstitialAd = true
    adUnitConfig.adConfiguration.isRewarded = true   // added
    adUnitConfig.adPosition = .fullScreen
    ...
}
```

This is a pre-existing bug, not new surface area, but it changes outgoing request payloads for every existing `RewardedVideoAdUnit` (Original API) integration and should be called out explicitly for review, since it's a behavior change to a widely-used class rather than purely additive.

### 3.5 GAM specifically

No new GAM-specific class is introduced for Original API. The demo app's `GAMOriginalAPIDisplayRewardedViewController` reuses `RewardedDisplayAdUnit` directly, then loads and presents through GAM's own `RewardedAd` — consistent with §2's point that Original API is ad-server-rendered. Separately, `BaseRewardedAdUnit` (the class backing GAM's Rendering API, event-handler-mediated rewarded flow — `RewardedAdUnit` + `GAMRewardedAdEventHandler`) is unaffected by this PR: it was already format-agnostic, since it renders through Prebid's own `InterstitialController` regardless of whether the winning bid is video or display, and already shipped a display-rewarded demo (`GAMDisplayRewardedViewController`) before PR #1274.

## 4. Surface-by-Surface Summary

| Integration surface | Public API touched | Change in PR #1274 | Rendering owner |
|---|---|---|---|
| Original API (`RewardedVideoAdUnit` / `RewardedDisplayAdUnit`) | New `RewardedDisplayAdUnit : InterstitialAdUnit`; `RewardedVideoAdUnit` `isRewarded` fix | New banner-format class, plus a fix for a missing `isRewarded` flag on the existing video class (§3.4) | Publisher/ad-server SDK — GAM in the Original API demo this PR adds. Prebid supplies the bid and targeting only. |
| Mediation API (`MediationRewardedAdUnit`; demoed with AdMob and MAX) | `adFormats` property + new `init(configId:mediationDelegate:adFormats:)` | `adFormats` made mutable; publisher can request `[.banner]`, `[.video]`, or both | Prebid SDK, via the Prebid-authored mediation adapter (e.g. `PrebidAdMobRewardedAdapter`, `PrebidMAXRewardedAdapter`) once Prebid's bid wins the ad network's mediation waterfall. The ad network's SDK still relays the reward callback to the publisher. |
| In-App Bidding / Multiformat (`PrebidAdUnit` + `PrebidRequest`) | `PrebidAdUnit.fetchDemand(...)` | Bug fix: `isInterstitialAd` now derived as `isInterstitial \|\| isRewarded`, so rewarded requests are always fullscreen even if `isInterstitial` wasn't set | Rendering is publisher-owned; this PR only fixes Prebid's request signaling. |
| GAM Rendering API (`RewardedAdUnit` / `BaseRewardedAdUnit` + `GAMRewardedAdEventHandler`) — pre-existing, not part of this PR | None | Already supported both rewarded video and rewarded display before PR #1274; included here for completeness since it's easy to assume this gap was SDK-wide | Prebid SDK — same `InterstitialController`-based rendering as the Mediation API row above. |

## 5. Design Rationale

- **Subclass `InterstitialAdUnit` rather than create a new class hierarchy:** `InterstitialAdUnit` already defaults to banner format, fullscreen position, and exposes both `bannerParameters` and `videoParameters`. `RewardedDisplayAdUnit` only needs to add one line (`isRewarded = true`) plus the same `minWidthPerc`/`minHeightPerc` convenience initializer `RewardedVideoAdUnit` already has. This keeps the two rewarded classes symmetric: `RewardedVideoAdUnit` extends `AdUnit` + `VideoBasedAdUnitProtocol` with `adFormats: [.video]` fixed; `RewardedDisplayAdUnit` extends `InterstitialAdUnit` with `adFormats` defaulting to `[.banner]`.
- **Expose `adFormats` as mutable on `MediationRewardedAdUnit` instead of adding a third rewarded class:** The Mediation API's underlying `AdUnitConfig` already supports arbitrary `adFormats` sets; the only missing piece was a public setter and a convenience initializer. This also gives multiformat rewarded (banner + video in one request) to Mediation API for free. Original API's dedicated classes (`RewardedVideoAdUnit`, `RewardedDisplayAdUnit`) remain single-format by design — multiformat rewarded is already reachable today via the in-app-bidding `PrebidAdUnit` + `PrebidRequest` path (see §6 and the open question in §9), so the gap is narrower than "Original API has no multiformat story at all."
- **Fix the fullscreen inference instead of asking publishers to also set `isInterstitial`:** Every other rewarded surface in the SDK (`RewardedVideoAdUnit`, `RewardedDisplayAdUnit`, `MediationRewardedAdUnit`, `BaseRewardedAdUnit`) treats rewarded as unconditionally fullscreen. Requiring `PrebidRequest(isRewarded: true, isInterstitial: true)` as two independent flags invites publisher error with no corresponding benefit — a rewarded ad that isn't fullscreen isn't a supported configuration anywhere else in the codebase.
- **Leave the reward-granting runtime (timer, close button, close action) untouched:** For the two surfaces where Prebid actually renders the creative — Mediation API (via the Prebid-authored adapters) and the pre-existing GAM Rendering API — that logic was already keyed off `adConfiguration.isRewarded` rather than creative format, so no changes were needed in `PBMHTMLCreative`, `PBMVideoCreative`, or `ModalViewController`. Original API doesn't have this runtime in scope, since the ad server SDK renders and grants the reward itself on that path (§2). The AdMob/MAX examples in this PR are Mediation API examples, not Original API. Scoping this PR to configuration-surface changes only keeps the diff reviewable and avoids touching well-tested, shared rendering code.

## 6. Alternatives Considered

| Alternative | Why not chosen | Reconsider if… |
|---|---|---|
| New `RewardedAdUnit` base class shared by video + display (format-agnostic), deprecating `RewardedVideoAdUnit` | Breaking change for every existing publisher on Original API; no functional gain since both already share `AdUnit`/`AdUnitConfig` underneath | The SDK later needs a third rewarded format (e.g. native) and the per-format subclass count becomes unwieldy |
| Separate `RewardedMultiformatAdUnit` class for Mediation API instead of a mutable `adFormats` setter | Adds a class whose only job is to hold a `Set<AdFormat>`; existing `MediationRewardedAdUnit` already has the config plumbing, just not exposed | Multiformat rewarded needs enough extra config (e.g. per-format parameters) that bolting it onto one class gets crowded |
| Lock `RewardedDisplayAdUnit.adFormats` to `[.banner]` (read-only override) | Matches `RewardedVideoAdUnit`'s spirit (one class = one format), but `InterstitialAdUnit` — its own superclass — already exposes a mutable `adFormats` today, so this would be a new restriction rather than consistency with existing precedent | Open question below — maintainers may prefer the stricter, harder-to-misuse version |
| Fix `isInterstitialAd` inference inside `AdUnitConfig` instead of `PrebidAdUnit.fetchDemand` | `AdUnitConfig` is shared by every ad unit type (banner, native, ...); baking a rewarded-implies-fullscreen rule in there risks side effects outside the multiformat flow | A second call site is found needing the same derivation, suggesting the rule belongs lower in the stack |

## 7. Backward Compatibility

- `RewardedDisplayAdUnit`, `MediationRewardedAdUnit.adFormats`, and the new `MediationRewardedAdUnit` initializer are purely additive — no existing call site changes behavior.
- The `PrebidAdUnit` fullscreen-inference fix (§3.3) only changes behavior for callers that explicitly set `isRewarded: true`; non-rewarded traffic is unaffected (`isInterstitial || false == isInterstitial`).
- **The `RewardedVideoAdUnit` fix (§3.4) is the one behavior change to already-shipped traffic:** existing Original API rewarded video integrations will start sending `imp.rwdd = 1` (and `ext.prebid.is_rewarded_inventory = true`) in the bid request where they previously didn't. This should surface Prebid Server-side rewarded-eligible demand that wasn't being matched before; we are not aware of a scenario where sending the correct signal reduces fill, but flagging for visibility since it's a request-payload change to existing, in-market integrations.

## 8. Testing

PR #1274 adds unit coverage for each surface:

- `RewardedDisplayAdUnit` creation (Swift + Obj-C) and the `minWidthPerc`/`minHeightPerc` convenience initializer.
- `RewardedVideoAdUnit` now asserts `isRewarded == true` alongside the existing `isInterstitialAd`/`adPosition`/`adFormats` assertions — a regression test guarding the outgoing-request signaling fix in §3.4.
- `PrebidAdUnit.fetchDemand(request:)` and `fetchDemand(adObject:request:)` both asserting `isInterstitialAd`, `isRewarded`, `adPosition`, and `adFormats` for a `PrebidRequest(bannerParameters:, isRewarded: true)`.
- `MediationRewardedAdUnit`: `adFormats` setter, the new `adFormats` initializer, multiformat (`[.banner, .video]`) configuration, and that `bannerParameters`/`videoParameters` both stay reachable after switching formats.
- Parameter-builder test asserting the outgoing OpenRTB imp for a display rewarded mediation ad unit carries banner (not video), rewarded (`imp.rwdd`) = 1, `instl` = 1, and `ext.prebid.is_rewarded_inventory` = true.

Demo/manual coverage: new AdMob, GAM Original API, and MAX "Display Rewarded" example view controllers were added to PrebidDemo (both Swift and Objective-C targets), covering the fetch → load → present flow per integration surface for manual verification — these are sample app controllers, not automated end-to-end tests.

## 9. Open Questions for Maintainers — iOS

- **Should `RewardedDisplayAdUnit` lock `adFormats` to `[.banner]`?** It currently inherits a mutable `adFormats` setter from `InterstitialAdUnit`, so a publisher could technically set `adFormats = [.video]` on a `RewardedDisplayAdUnit` instance, which would be semantically confusing (a "display" ad unit requesting video). Our recommendation: keep the inherited mutability rather than lock it down. `InterstitialAdUnit` itself already exposes the same mutable `adFormats` to every one of its subclasses, so locking it here would be a new restriction, not a return to an existing norm, and it would only guard against a self-inflicted misconfiguration. We'd instead document that `RewardedDisplayAdUnit` is intended for `.banner`, and point publishers who want mixed formats through Original API to `PrebidAdUnit` + `PrebidRequest` (§3.3/§5 above) or, if that's insufficient, a future dedicated multiformat rewarded class (see the question below). Open to the stricter get-only version if maintainers weigh the misuse risk differently.
- **Should Original API get a dedicated multiformat rewarded class, analogous to Mediation API's `adFormats` set?** Multiformat rewarded (banner + video in one auction) is already reachable today via `PrebidAdUnit` + `PrebidRequest` (in-app bidding) and via `MediationRewardedAdUnit`, so this isn't a hard gap for publishers who can use either of those paths. The open question is narrower: whether publishers specifically on Original API's dedicated classes (`RewardedVideoAdUnit` / `RewardedDisplayAdUnit`) — as opposed to in-app bidding — should get an equivalent, which would want a third class (e.g. `RewardedAdUnit` accepting an `adFormats` set) rather than extending the two single-format classes further. Out of scope for this PR but worth deciding now given the naming precedent it sets.
- **Is the `RewardedVideoAdUnit` request-payload change (§3.4) acceptable to ship without a feature flag or migration note?** It's a bug fix (declaring a flag that was always intended to be true), but it does change outgoing request bytes for all existing Original API rewarded video integrations. We propose calling it out in the release notes rather than gating it, since withholding the correct rewarded signal is strictly worse for publishers than sending it.
- **Naming: "`RewardedDisplayAdUnit`" vs. "`RewardedBannerAdUnit`":** The SDK's `AdFormat` case is `.banner`, and `BannerAdUnit` is the non-rewarded analog for static creative, but issue #713 and the ad-industry convention both use "Display" for this format family. We went with `RewardedDisplayAdUnit` for consistency with the issue title; happy to rename if the community prefers alignment with the existing `BannerAdUnit`/`.banner` naming instead.

---

# PART II — ANDROID IMPLEMENTATION (PR #963)

## Problem Statement (Android)

PR #963 links back to iOS issue #713 rather than a separate Android issue, since the underlying gap — no supported way to request rewarded display inventory — applies to both platforms. But Android's starting point wasn't identical to iOS's: Android's Original API and Mediation API had the same missing-format gap iOS did, while its GAM Rendering API / in-app-bidding class (`RewardedAdUnit`) had a narrower problem — it already requested both formats under the hood, but offered publishers no way to request just one, and its AdMob adapter silently assumed every rewarded creative was video. So on Android, this proposal closes:

- The same Original API and Mediation API format gaps as iOS (no way to request banner-only rewarded through `RewardedVideoAdUnit` or the old `MediationRewardedVideoAdUnit`).
- A configurability gap in `RewardedAdUnit` (GAM Rendering API / in-app bidding): it always requested both formats with no way to narrow to one.
- A latent correctness bug in the AdMob mediation adapter that assumed every rewarded creative was video (§11.4) — found while implementing this feature, not something issue #713 called out directly.

## 10. Background: What Already Exists (Android)

Android's rewarded surfaces map onto three integration paths rather than iOS's four, because Android's Rendering API class already does double duty as both the GAM-mediated path and the standalone in-app-bidding path:

- **Original API (`RewardedVideoAdUnit`, and now `RewardedDisplayAdUnit`) is ad-server-rendered, not Prebid-rendered — same as iOS.** Prebid's role is limited to `fetchDemand(...)`; GAM's own native `RewardedAd` format (see `GamOriginalApiDisplayRewardedActivity.kt`) renders the creative and grants the reward through its own callback. `RewardedVideoAdUnit` (`org.prebid.mobile.RewardedVideoAdUnit`) is untouched by this PR — unlike iOS, it already called `configuration.setRewarded(true)` in its constructor, so there was no equivalent request-signaling bug to fix here.
- **Mediation API (`MediationRewardedAdUnit`) and GAM Rendering API / in-app bidding (`RewardedAdUnit`) are Prebid-rendered.** When Prebid's bid wins the mediation waterfall (via `PrebidRewardedAdapter` for AdMob or `MaxRewardedManager` for MAX) or the GAM line item (via `GamRewardedEventHandler`) or is used standalone (via `StandaloneRewardedVideoEventHandler` for in-app bidding), Prebid's own `InterstitialController` renders the creative. Unlike iOS, where the GAM Rendering API path already supported both formats before this PR, Android's `RewardedAdUnit` previously requested a hardcoded `EnumSet.of(AdFormat.INTERSTITIAL, AdFormat.VAST)` with no way to request just one format — so this PR had to add configurability here, not just reuse it.
- **A real bug was found on the rendering side, not the request side.** The AdMob mediation adapter (`PrebidRewardedAdapter`) unconditionally called AdMob's `onVideoStart()` and `onVideoComplete()` callbacks whenever a rewarded ad displayed or closed — correct for rewarded video, wrong for a display creative that never played a video. This proposal fixes it by adding `InterstitialController.getAdUnitIdentifierType()` so the adapter can check the winning bid's actual format before firing those callbacks (§11.4).

Because format-aware rendering already existed at the Prebid-rendered layer (same `InterstitialController` class as iOS's equivalent), this PR is similarly narrow in intent, but broader in one respect: the GAM Rendering API / in-app-bidding surface needed new constructor overloads here, whereas the equivalent iOS surface needed none.

## 11. Proposed API (Android)

### 11.1 Original API — new `RewardedDisplayAdUnit`

A new public class, sibling to `RewardedVideoAdUnit`, subclassing `InterstitialAdUnit`:

```java
public class RewardedDisplayAdUnit extends InterstitialAdUnit {

    public RewardedDisplayAdUnit(@NonNull String configId) {
        super(configId);
        configuration.setRewarded(true);
        configuration.setAdPosition(AdPosition.FULLSCREEN);
    }

    public RewardedDisplayAdUnit(
            @NonNull String configId,
            @IntRange(from = 0, to = 100) int minWidthPerc,
            @IntRange(from = 0, to = 100) int minHeightPerc
    ) {
        this(configId);
        configuration.setMinSizePercentage(new AdSize(minWidthPerc, minHeightPerc));
    }
}
```

Same shape as iOS's `RewardedDisplayAdUnit`, down to the `minWidthPerc`/`minHeightPerc` convenience constructor. The explicit `setAdPosition(FULLSCREEN)` call is redundant with `InterstitialAdUnit`'s default, but kept to make the rewarded fullscreen invariant visible at the class boundary.

### 11.2 Mediation API — new `MediationRewardedAdUnit`, old class deprecated

Unlike iOS, which made the existing `MediationRewardedAdUnit`'s `adFormats` mutable, Android introduces a new `MediationRewardedAdUnit` class and deprecates the old video-only one on top of it:

```java
public class MediationRewardedAdUnit extends MediationBaseFullScreenAdUnit {

    // Defaults to video-only, for source compatibility
    public MediationRewardedAdUnit(
        Context context, String configId, PrebidMediationDelegate mediationDelegate
    ) {
        this(context, configId, EnumSet.of(AdUnitFormat.VIDEO), mediationDelegate);
    }

    // Use EnumSet.of(AdUnitFormat.BANNER) for display rewarded,
    // .VIDEO for video, or both for multiformat rewarded.
    public MediationRewardedAdUnit(
        Context context, String configId,
        @NonNull EnumSet<AdUnitFormat> adUnitFormats,
        PrebidMediationDelegate mediationDelegate
    ) {
        super(context, configId, null, mediationDelegate);
        adUnitConfig.setAdUnitFormats(adUnitFormats);
    }
}

/**
 * @deprecated Use MediationRewardedAdUnit. This class remains
 * as a video-only convenience wrapper.
 */
@Deprecated
public class MediationRewardedVideoAdUnit extends MediationRewardedAdUnit {
    public MediationRewardedVideoAdUnit(
        Context context, String configId, PrebidMediationDelegate mediationDelegate
    ) {
        super(context, configId, mediationDelegate);
    }
}
```

`MediationBaseFullScreenAdUnit` (the shared superclass) gained public `setBannerParameters`/`getBannerParameters` and `setVideoParameters`/`getVideoParameters` accessors in this PR, so both parameter sets stay reachable regardless of the current `adUnitFormats` — the same guarantee iOS's §3.2 provides.

### 11.3 GAM Rendering API / In-App Bidding — `RewardedAdUnit` format-aware constructors

`RewardedAdUnit` (`org.prebid.mobile.api.rendering.RewardedAdUnit`) previously had no way to request a single format — every constructor hardcoded `EnumSet.of(AdFormat.INTERSTITIAL, AdFormat.VAST)` internally. This PR adds format-aware overloads for both the standalone (in-app bidding) and GAM-backed constructors:

```java
// Standalone (in-app bidding) — defaults to both formats, unchanged behavior
public RewardedAdUnit(Context context, String configId) {
    this(context, configId, EnumSet.of(AdUnitFormat.BANNER, AdUnitFormat.VIDEO),
         new StandaloneRewardedVideoEventHandler());
}

// Standalone (in-app bidding) — caller-specified formats
public RewardedAdUnit(Context context, String configId, EnumSet<AdUnitFormat> adUnitFormats) {
    this(context, configId, adUnitFormats, new StandaloneRewardedVideoEventHandler());
}

// GAM-backed (or other RewardedEventHandler) — caller-specified formats
public RewardedAdUnit(
    Context context, String configId,
    EnumSet<AdUnitFormat> adUnitFormats, RewardedEventHandler eventHandler
) {
    ...
    adUnitConfiguration.setAdUnitFormats(adUnitFormats);
    adUnitConfiguration.setRewarded(true);
    ...
}
```

Both the GAM Rendering API demo (`GamRenderingApiDisplayRewardedActivity`, passing `GamRewardedEventHandler`) and the standalone in-app-bidding demo (`InAppDisplayRewardedActivity`, using the no-event-handler constructor) are the same `RewardedAdUnit` class — they differ only in which `RewardedEventHandler` is supplied. This is the key structural difference from iOS, where GAM Rendering API and in-app bidding are two unrelated class hierarchies (§17).

### 11.4 AdMob mediation adapter fix — no more spurious video callbacks for display creatives

`PrebidRewardedAdapter` (the AdMob custom-event adapter) previously called AdMob's video-lifecycle callbacks unconditionally:

```java
// before
public void onInterstitialDisplayed() {
    rewardedAdCallback.reportAdImpression();
    rewardedAdCallback.onAdOpened();
    rewardedAdCallback.onVideoStart();  // fires even for a display creative
}
public void onInterstitialClosed() {
    rewardedAdCallback.onVideoComplete();  // fires even for a display creative
    rewardedAdCallback.onAdClosed();
}
```

This proposal adds a getter to `InterstitialController` so the adapter can check the winning bid's actual format, and gates the video-specific callbacks on it:

```java
// InterstitialController.java — new getter
@Nullable
public AdFormat getAdUnitIdentifierType() {
    return adUnitIdentifierType;
}

// PrebidRewardedAdapter.java — after
private boolean isLoadedAdVideo() {
    return interstitialController != null
        && interstitialController.getAdUnitIdentifierType() == AdFormat.VAST;
}
void notifyAdDisplayed() {
    rewardedAdCallback.reportAdImpression();
    rewardedAdCallback.onAdOpened();
    if (isLoadedAdVideo()) rewardedAdCallback.onVideoStart();
}
void notifyAdClosed() {
    if (isLoadedAdVideo()) rewardedAdCallback.onVideoComplete();
    rewardedAdCallback.onAdClosed();
}
```

This is Android's analog to iOS's §3.4 collateral fix: both platforms had an assumption baked in from when only rewarded video existed, just in different places — iOS's was on the outgoing request (missing `isRewarded` signal), Android's is on the incoming response/callback wiring (video-only lifecycle events). MAX's adapter (`MaxRewardedManager`) needed no equivalent change; its rewarded callback interface doesn't have AdMob's separate `onVideoStart`/`onVideoComplete` split.

## 12. Surface-by-Surface Summary (Android)

| Integration surface | Public API touched | Change in PR #963 | Rendering owner |
|---|---|---|---|
| Original API (`RewardedVideoAdUnit` / `RewardedDisplayAdUnit`) | New `RewardedDisplayAdUnit extends InterstitialAdUnit` | New banner-format class. `RewardedVideoAdUnit` was already correct (already set `isRewarded` on the request) — no collateral fix needed here, unlike iOS §3.4. | Publisher/ad-server SDK — GAM in the Original API demo this PR adds. Prebid supplies the bid and targeting only. |
| Mediation API (new `MediationRewardedAdUnit`; old `MediationRewardedVideoAdUnit` deprecated) | New `MediationRewardedAdUnit(context, configId, EnumSet<AdUnitFormat>, delegate)`; `MediationRewardedVideoAdUnit` now extends it as a video-only, `@Deprecated` wrapper | New formats-aware class; deprecates and re-implements the old class on top of it, rather than adding a mutable property to the existing class | Prebid SDK, via the Prebid-authored mediation adapter (`PrebidRewardedAdapter` for AdMob, `MaxRewardedManager` for MAX) once Prebid's bid wins the ad network's mediation waterfall. |
| GAM Rendering API / In-App Bidding (`RewardedAdUnit`, both `StandaloneRewardedVideoEventHandler` and `GamRewardedEventHandler`) | New `RewardedAdUnit(context, configId, EnumSet<AdUnitFormat>[, eventHandler])` constructor overloads | Previously hardcoded to request `EnumSet.of(AdFormat.INTERSTITIAL, AdFormat.VAST)` (i.e. always both); now the formats are caller-specified via the public `AdUnitFormat` enum, with both-formats as the still-unchanged default | Prebid SDK — same `InterstitialController`-based rendering used by Mediation API. |
| AdMob mediation adapter (`PrebidRewardedAdapter`) | New `InterstitialController.getAdUnitIdentifierType()` getter | Fixes a bug: the adapter unconditionally fired AdMob's `onVideoStart()`/`onVideoComplete()` callbacks on display/open/close, even for a non-video creative; now gated on the winning bid's actual format | Prebid SDK (adapter-internal correctness fix, not a new surface) |

## 13. Design Rationale (Android)

- **Deprecate-and-wrap instead of making the existing class format-mutable:** iOS extended its existing `MediationRewardedAdUnit` in place because Swift call sites read a settable `adFormats` property cleanly. Android instead introduced a new formats-aware `MediationRewardedAdUnit` and turned the old `MediationRewardedVideoAdUnit` into a thin `@Deprecated` subclass. This keeps every existing Android call site source-compatible (the deprecated class still compiles and behaves identically) while giving the deprecation warning a clear migration target — arguably clearer for Java/Kotlin callers than a mutable property would have been, since the format is fixed at construction rather than settable after the fact.
- **Add format-aware constructors to `RewardedAdUnit` rather than a new class:** `RewardedAdUnit` already serves two roles (GAM-backed and standalone in-app bidding) via which `RewardedEventHandler` is passed in; adding `EnumSet<AdUnitFormat>` constructor overloads extends both roles at once without a new class per role, mirroring how iOS added formats to the existing `MediationRewardedAdUnit` rather than forking it.
- **Default `RewardedAdUnit` to both formats rather than video-only:** The pre-existing hardcoded `EnumSet.of(AdFormat.INTERSTITIAL, AdFormat.VAST)` already requested both formats for every publisher, whether or not they used the display side. Defaulting the new `EnumSet.of(AdUnitFormat.BANNER, AdUnitFormat.VIDEO)` constructor the same way preserves that existing behavior exactly — no existing `RewardedAdUnit` publisher sees a narrower request than before.
- **Fix the AdMob adapter's video-callback assumption at the adapter layer, not in `InterstitialController` itself:** The bug is specific to AdMob's callback vocabulary (which splits `reportAdImpression`/`onAdOpened` from `onVideoStart`/`onVideoComplete`); `InterstitialController.getAdUnitIdentifierType()` is a small, generic getter that any adapter can use to make its own format-appropriate decision, rather than baking AdMob-specific behavior into shared rendering code.

## 14. Backward Compatibility (Android)

- `RewardedDisplayAdUnit` and the new `RewardedAdUnit`/`MediationRewardedAdUnit` constructor overloads are purely additive.
- `MediationRewardedVideoAdUnit` remains usable and behaviorally identical — it's now `@Deprecated` and implemented as a subclass, but every existing call site still compiles and still requests video-only rewarded.
- `RewardedAdUnit`'s default (no-format) constructors keep requesting both formats, matching pre-PR behavior exactly — no publisher sees a change unless they opt into the new `EnumSet<AdUnitFormat>` constructor.
- **The AdMob adapter fix (§11.4) is the one behavior change to already-shipped traffic:** any existing AdMob rewarded-video integration that was somehow receiving a non-video winning bid (which shouldn't happen given the old class only requested VAST) would have seen incorrect `onVideoStart`/`onVideoComplete` firing; this fix only removes an incorrect callback, it doesn't add a new one video integrations didn't already get.

## 15. Testing (Android)

PR #963 adds unit coverage (Robolectric/JUnit) for each surface:

- `RewardedDisplayAdUnit` creation and the `minWidthPerc`/`minHeightPerc` convenience constructor (`AdUnitSuccessorTest`), plus an added `isRewarded` assertion on the pre-existing `RewardedVideoAdUnit` test — a regression guard, not a fix, since that class was already correct.
- `MediationRewardedAdUnit`: format-to-`AdFormat` mapping for BANNER, VIDEO, and BANNER+VIDEO, the video-only default constructor, `isRewarded`/`AdPosition.FULLSCREEN` assertions, and that `bannerParameters`/`videoParameters` stay reachable after a format change (`MediationRewardedAdUnitTest` — new file).
- `RewardedAdUnit`: the same format-to-`AdFormat` mapping for the standalone/GAM-backed constructors (`RewardedAdUnitTest`).
- Parameter-builder test asserting the outgoing OpenRTB imp for a display rewarded ad unit carries a banner (not video), `imp.rewarded = 1`, `imp.instl = 1`, `imp.rwdd = 1`, and `ext.prebid.is_rewarded_inventory = 1` (`BasicParameterBuilderTest.displayRewarded_BannerImpWithRewardedSignals`) — confirms the same `imp.rwdd` wire signal as iOS.

Demo/manual coverage: new GAM Original API, GAM Rendering API, AdMob, AppLovin MAX, and standalone in-app-bidding "Display Rewarded" example activities were added to both PrebidDemoJava and PrebidDemoKotlin — sample-app activities for manual verification, not automated end-to-end tests.

## 16. Open Questions for Maintainers — Android

- **Is deprecate-and-wrap (`MediationRewardedVideoAdUnit` → `MediationRewardedAdUnit`) the right pattern, or should Android instead have made the existing class format-mutable like iOS did?** The deprecate-and-wrap approach is arguably cleaner for a construction-time-fixed format, but it's a different migration shape than iOS's in-place mutable property — see §17/§18 for whether these should converge.
- **Should `RewardedAdUnit`'s default (no-format) constructors keep defaulting to both formats, or should that default become video-only to match `RewardedVideoAdUnit`'s historical behavior more narrowly?** We kept the existing both-formats default to avoid any behavior change for current callers, but it does mean an unconfigured `RewardedAdUnit` today silently remains eligible for display inventory, even for publishers who mentally modeled this class as rewarded-video-first.
- **Does the AdMob adapter fix (§11.4) need a changelog callout even though it only removes an incorrect callback?** We believe no existing integration could have been relying on the incorrect behavior (a video-only class couldn't have surfaced a non-video winning bid before this PR), but flagging for visibility since it touches adapter code already in production.

---

# PART III — CROSS-PLATFORM CONSISTENCY

## 17. Cross-Platform Comparison

Both platforms solve the same problem (issue #713) with the same overall shape — a new `*DisplayAdUnit` for Original API, format-awareness added to the Mediation API class, and no changes to the shared reward-granting runtime — but they diverge in several concrete ways worth reconciling before both PRs merge:

| Concept | iOS (PR #1274) | Android (PR #963) |
|---|---|---|
| Original API display class | `RewardedDisplayAdUnit : InterstitialAdUnit` | `RewardedDisplayAdUnit extends InterstitialAdUnit` — same shape, same name |
| Mediation API: how format choice was added | Existing `MediationRewardedAdUnit` made `adFormats` mutable; no deprecation | New `MediationRewardedAdUnit` added; old `MediationRewardedVideoAdUnit` deprecated and rewritten as a wrapper around it |
| Publisher-facing format type | `Set<AdFormat>` (`.banner` / `.video`) — the same `AdFormat` type used everywhere in the SDK | `EnumSet<AdUnitFormat>` (`BANNER` / `VIDEO`) — a separate public type from the SDK-internal `AdFormat` (`INTERSTITIAL` / `VAST`) used in `AdUnitConfig` |
| Collateral bug found while implementing this feature | `RewardedVideoAdUnit` (Original API) never set `isRewarded` on the request — request-signaling bug, fixed in `AdUnitConfig` construction | `PrebidRewardedAdapter` (AdMob) unconditionally fired video-lifecycle callbacks for any creative — response-handling bug, fixed via a new format getter on `InterstitialController` |
| "GAM Rendering API" / "in-app bidding" surface | Two separate class hierarchies: `PrebidAdUnit` + `PrebidRequest` (in-app bidding) vs. `RewardedAdUnit` + `BaseRewardedAdUnit` + `GAMRewardedAdEventHandler` (GAM Rendering API). The GAM Rendering API one already supported display before this PR. | One `RewardedAdUnit` class serves both roles, distinguished only by which `RewardedEventHandler` is passed (`StandaloneRewardedVideoEventHandler` vs. `GamRewardedEventHandler`). Neither role supported configurable formats before this PR. |
| OpenRTB wire signal | `imp.rwdd` (property name `rewarded`), plus legacy `ext.prebid.is_rewarded_inventory` | Same: `imp.rwdd`, plus `ext.prebid.is_rewarded_inventory` — confirmed identical on the wire |

None of these differences are functionally wrong on either platform individually; they're the kind of divergence that accumulates when two platform teams solve the same problem in parallel without a shared spec — which is exactly why this document now covers both.

## 18. Cross-Platform Open Questions

- **Should the Mediation API migration pattern converge?** iOS made an existing class's `adFormats` mutable; Android added a new class and deprecated the old one. Picking one pattern (likely case-by-case, since Swift's property-mutation idiom and Java's deprecate-and-wrap idiom are both normal for their respective ecosystems) at least means future format-adding PRs on either platform don't have to re-litigate this.
- **Should the publisher-facing format type be named consistently?** iOS publishers write `Set<AdFormat>` with cases `.banner`/`.video` everywhere. Android publishers write `EnumSet<AdUnitFormat>` with cases `BANNER`/`VIDEO` for this feature, but the SDK-internal `AdUnitConfig` on Android uses a different `AdFormat` enum (`INTERSTITIAL`/`VAST`) for the same concept. A publisher reading both platforms' docs side by side will reasonably wonder whether `AdFormat` and `AdUnitFormat` mean the same thing (they don't, on Android) or different things (they do, in effect, match iOS's `AdFormat`). Worth deciding whether Android's public surface should rename toward `AdFormat` for cross-platform consistency, independent of this PR's scope.
- **Should "GAM Rendering API" and "in-app bidding" be unified on iOS the way they already are on Android?** Android's single `RewardedAdUnit` class (parameterized by event handler) versus iOS's two separate hierarchies (`PrebidAdUnit`+`PrebidRequest` vs. `RewardedAdUnit`+`BaseRewardedAdUnit`+`GAMRewardedAdEventHandler`) is a pre-existing architectural difference, not something introduced by either rewarded-display PR — out of scope to fix here, but worth a maintainer decision on which shape is the long-term target so future features don't have to solve this same problem on both platforms independently again.
- **Should both platforms' collateral bugs (§3.4 iOS, §11.4 Android) be called out together in one release-notes entry?** They're unrelated bugs in unrelated code, found independently while building the same feature on two platforms, but a publisher shipping both SDK versions together may reasonably want a single "rewarded ad fixes in this release" note rather than two separate, differently-worded ones.

**Comment by YuriyVelichkoPI** [comment]
The doc version of proposal to discuss, comemnt, and edit
https://docs.google.com/document/d/1SJN6J2njBNcThxTlUTQ3mmHjwwxn0n7lVPNde7l3lOE/edit?usp=sharing
